File: crates/spanda-core/src/deploy_remote.rs
//! Remote OTA rollout via HTTP deploy agents.

use crate::deploy_http::{http_request, parse_http_url, HttpResponse};
use crate::deploy_service::{
    deploy_target_key, DeployPlan, RolloutOptions, RolloutResult, RolloutStep, RolloutStepStatus,
    RolloutStrategy,
};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fs;
use std::path::{Path, PathBuf};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct DeployAgentEntry {
    pub target: String,
    pub url: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub token: Option<String>,
}

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct DeployAgentRegistry {
    pub agents: Vec<DeployAgentEntry>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct AgentStatusResponse {
    pub ok: bool,
    pub target: String,
    pub current_version: String,
    pub previous_version: Option<String>,
    pub healthy: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
struct RolloutRequest {
    target: String,
    version: String,
    program: Option<String>,
    program_hash: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct AgentRolloutResponse {
    pub ok: bool,
    pub target: String,
    pub version: String,
    pub previous_version: Option<String>,
    #[serde(default)]
    pub error: Option<String>,
}

pub fn default_agents_path() -> PathBuf {
    PathBuf::from(".spanda/deploy-agents.json")
}

pub fn load_agent_registry(path: &Path) -> DeployAgentRegistry {
    if !path.exists() {
        return DeployAgentRegistry::default();
    }
    fs::read_to_string(path)
        .ok()
        .and_then(|text| serde_json::from_str(&text).ok())
        .unwrap_or_default()
}

pub fn save_agent_registry(path: &Path, registry: &DeployAgentRegistry) -> Result<(), String> {
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent).map_err(|e| e.to_string())?;
    }
    let text = serde_json::to_string_pretty(registry).map_err(|e| e.to_string())?;
    fs::write(path, text).map_err(|e| e.to_string())
}

pub fn register_agent(
    registry: &mut DeployAgentRegistry,
    target: String,
    url: String,
    token: Option<String>,
) -> Result<(), String> {
    // Validate URL shape before recording the agent endpoint.
    parse_http_url(&url)?;
    registry.agents.retain(|entry| entry.target != target);
    registry.agents.push(DeployAgentEntry { target, url, token });
    registry.agents.sort_by(|a, b| a.target.cmp(&b.target));
    Ok(())
}

pub fn lookup_agent<'a>(
    registry: &'a DeployAgentRegistry,
    target: &str,
) -> Option<&'a DeployAgentEntry> {
    registry.agents.iter().find(|entry| entry.target == target)
}

fn agent_endpoint(base_url: &str, path: &str) -> Result<String, String> {
    let parsed = parse_http_url(base_url)?;
    Ok(format!(
        "{}://{}:{}{}",
        parsed.scheme, parsed.host, parsed.port, path
    ))
}

fn decode_response<T: for<'de> Deserialize<'de>>(response: HttpResponse) -> Result<T, String> {
    if response.status >= 400 {
        return Err(format!("agent HTTP {}: {}", response.status, response.body));
    }
    serde_json::from_str(&response.body).map_err(|e| format!("invalid agent JSON: {e}"))
}

pub fn agent_health(entry: &DeployAgentEntry) -> Result<bool, String> {
    let url = agent_endpoint(&entry.url, "/v1/health")?;
    let response = http_request("GET", &url, None, entry.token.as_deref())?;
    let body: serde_json::Value = decode_response(response)?;
    Ok(body.get("ok").and_then(|v| v.as_bool()).unwrap_or(false))
}

pub fn agent_status(entry: &DeployAgentEntry) -> Result<AgentStatusResponse, String> {
    let url = agent_endpoint(&entry.url, "/v1/status")?;
    let response = http_request("GET", &url, None, entry.token.as_deref())?;
    decode_response(response)
}

pub fn agent_rollout(
    entry: &DeployAgentEntry,
    version: &str,
    program: Option<&str>,
    program_hash: Option<&str>,
) -> Result<AgentRolloutResponse, String> {
    let url = agent_endpoint(&entry.url, "/v1/rollout")?;
    let payload = serde_json::to_string(&RolloutRequest {
        target: entry.target.clone(),
        version: version.to_string(),
        program: program.map(str::to_string),
        program_hash: program_hash.map(str::to_string),
    })
    .map_err(|e| e.to_string())?;
    let response = http_request("POST", &url, Some(&payload), entry.token.as_deref())?;
    decode_response(response)
}

pub fn agent_rollback(entry: &DeployAgentEntry) -> Result<AgentRolloutResponse, String> {
    let url = agent_endpoint(&entry.url, "/v1/rollback")?;
    let payload = serde_json::to_string(&serde_json::json!({
        "target": entry.target,
    }))
    .map_err(|e| e.to_string())?;
    let response = http_request("POST", &url, Some(&payload), entry.token.as_deref())?;
    decode_response(response)
}

pub fn execute_remote_rollout(
    plan: &DeployPlan,
    options: &RolloutOptions,
    registry: &DeployAgentRegistry,
) -> RolloutResult {
    // Plan rollout steps locally, then push updates to registered deploy agents.
    let local = crate::deploy_service::plan_rollout(plan, options);
    if options.dry_run {
        return local;
    }

    let mut steps = Vec::new();
    let mut success = true;
    for step in &local.steps {
        if step.status == RolloutStepStatus::Skipped {
            steps.push(step.clone());
            continue;
        }
        let key = deploy_target_key(&step.robot_name, &step.hardware);
        let Some(agent) = lookup_agent(registry, &key) else {
            steps.push(RolloutStep {
                status: RolloutStepStatus::Failed,
                ..step.clone()
            });
            success = false;
            continue;
        };
        match agent_rollout(
            agent,
            &step.version,
            Some(&plan.program),
            plan.program_hash.as_deref(),
        ) {
            Ok(resp) if resp.ok => steps.push(RolloutStep {
                status: RolloutStepStatus::Deployed,
                ..step.clone()
            }),
            Ok(resp) => {
                steps.push(RolloutStep {
                    status: RolloutStepStatus::Failed,
                    version: resp.version,
                    ..step.clone()
                });
                success = false;
            }
            Err(_) => {
                steps.push(RolloutStep {
                    status: RolloutStepStatus::Failed,
                    ..step.clone()
                });
                success = false;
            }
        }
    }

    RolloutResult {
        strategy: options.strategy,
        version: options.version.clone(),
        dry_run: false,
        steps,
        success,
    }
}

pub fn execute_remote_rollback(
    plan: &DeployPlan,
    registry: &DeployAgentRegistry,
) -> RolloutResult {
    let mut steps = Vec::new();
    let mut success = false;
    for assignment in &plan.assignments {
        let key = deploy_target_key(&assignment.robot_name, &assignment.hardware);
        let step = RolloutStep {
            robot_name: assignment.robot_name.clone(),
            hardware: assignment.hardware.clone(),
            status: RolloutStepStatus::Skipped,
            version: "unknown".into(),
            phase_percent: None,
        };
        let Some(agent) = lookup_agent(registry, &key) else {
            steps.push(step);
            continue;
        };
        match agent_rollback(agent) {
            Ok(resp) if resp.ok => {
                success = true;
                steps.push(RolloutStep {
                    status: RolloutStepStatus::RolledBack,
                    version: resp.version,
                    ..step
                });
            }
            Ok(resp) => steps.push(RolloutStep {
                status: RolloutStepStatus::Failed,
                version: resp.version,
                ..step
            }),
            Err(_) => steps.push(RolloutStep {
                status: RolloutStepStatus::Failed,
                ..step
            }),
        }
    }
    RolloutResult {
        strategy: RolloutStrategy::All,
        version: "rollback".into(),
        dry_run: false,
        steps,
        success,
    }
}

pub fn missing_remote_targets(
    plan: &DeployPlan,
    registry: &DeployAgentRegistry,
) -> Vec<String> {
    let mut missing = Vec::new();
    for assignment in &plan.assignments {
        let key = deploy_target_key(&assignment.robot_name, &assignment.hardware);
        if lookup_agent(registry, &key).is_none() {
            missing.push(key);
        }
    }
    missing
}

pub fn registry_by_target(registry: &DeployAgentRegistry) -> HashMap<String, DeployAgentEntry> {
    registry
        .agents
        .iter()
        .cloned()
        .map(|entry| (entry.target.clone(), entry))
        .collect()
}

File: src/deploy-remote.ts
/**
 * Remote OTA deploy agent client and registry.
 * @module
 */

import { readFileSync, writeFileSync, mkdirSync, existsSync } from "node:fs";
import { dirname, resolve } from "node:path";
import {
  applyRollout,
  deployTargetKey,
  planRollout,
  type DeployPlan,
  type RolloutOptions,
  type RolloutResult,
  type RolloutStep,
  type CertificationProofSummary,
} from "./deploy-service.js";
import type { DeployArtifactBundle } from "./deploy-bundle.js";
import { remoteFetch } from "./http-fetch.js";

function certificationProofPayload(
  proof?: CertificationProofSummary,
): Record<string, unknown> | undefined {
  // Description:
  //     CertificationProofPayload.
  //
  // Inputs:
  //     proof?: CertificationProofSummary
  //         Caller-supplied proof?.
  //
  // Outputs:
  //     result: Record<string, unknown> | undefined
  //         Return value from `certificationProofPayload`.
  //
  // Example:

  //     const result = certificationProofPayload(proof?);

  if (!proof) return undefined;
  return {
    passed: proof.passed,
    passed_strict: proof.passedStrict,
    summary: proof.summary,
    error_count: proof.errorCount,
    warning_count: proof.warningCount,
  };
}

export type DeployAgentEntry = {
  target: string;
  url: string;
  token?: string;
};

export type DeployAgentRegistry = {
  agents: DeployAgentEntry[];
};

export type AgentStatusResponse = {
  ok: boolean;
  target: string;
  currentVersion: string;
  previousVersion?: string;
  healthy: boolean;
};

export function defaultAgentsPath(): string {
  // Description:
  //     DefaultAgentsPath.
  //
  // Inputs:
  //     None.
  //
  // Outputs:
  //     result: string
  //         Return value from `defaultAgentsPath`.
  //
  // Example:

  //     const result = defaultAgentsPath();

  return process.env.SPANDA_DEPLOY_AGENTS ?? ".spanda/deploy-agents.json";
}

export function emptyAgentRegistry(): DeployAgentRegistry {
  // Description:
  //     EmptyAgentRegistry.
  //
  // Inputs:
  //     None.
  //
  // Outputs:
  //     result: DeployAgentRegistry
  //         Return value from `emptyAgentRegistry`.
  //
  // Example:

  //     const result = emptyAgentRegistry();

  return { agents: [] };
}

export function loadAgentRegistry(text: string | null): DeployAgentRegistry {
  // Description:
  //     LoadAgentRegistry.
  //
  // Inputs:
  //     text: string | null
  //         Caller-supplied text.
  //
  // Outputs:
  //     result: DeployAgentRegistry
  //         Return value from `loadAgentRegistry`.
  //
  // Example:

  //     const result = loadAgentRegistry(text);

  if (!text) return emptyAgentRegistry();
  try {
    const parsed = JSON.parse(text) as DeployAgentRegistry;
    return { agents: parsed.agents ?? [] };
  } catch {
    return emptyAgentRegistry();
  }
}

export function serializeAgentRegistry(registry: DeployAgentRegistry): string {
  // Description:
  //     SerializeAgentRegistry.
  //
  // Inputs:
  //     registry: DeployAgentRegistry
  //         Caller-supplied registry.
  //
  // Outputs:
  //     result: string
  //         Return value from `serializeAgentRegistry`.
  //
  // Example:

  //     const result = serializeAgentRegistry(registry);

  return JSON.stringify(registry, null, 2);
}

export function readAgentRegistryFromDisk(path = defaultAgentsPath()): DeployAgentRegistry {
  // Description:
  //     ReadAgentRegistryFromDisk.
  //
  // Inputs:
  //     path = defaultAgentsPath(): input value
  //         Caller-supplied path = defaultAgentsPath().
  //
  // Outputs:
  //     result: DeployAgentRegistry
  //         Return value from `readAgentRegistryFromDisk`.
  //
  // Example:

  //     const result = readAgentRegistryFromDisk(path = defaultAgentsPath());

  if (!existsSync(path)) return emptyAgentRegistry();
  return loadAgentRegistry(readFileSync(path, "utf-8"));
}

export function writeAgentRegistryToDisk(registry: DeployAgentRegistry, path = defaultAgentsPath()): void {
  // Description:
  //     WriteAgentRegistryToDisk.
  //
  // Inputs:
  //     registry: DeployAgentRegistry
  //         Caller-supplied registry.
  //     path = defaultAgentsPath(): input value
  //         Caller-supplied path = defaultAgentsPath().
  //
  // Outputs:
  //     None.
  //
  // Example:

  //     const result = writeAgentRegistryToDisk(registry, path = defaultAgentsPath());

  const abs = resolve(path);
  mkdirSync(dirname(abs), { recursive: true });
  writeFileSync(abs, serializeAgentRegistry(registry));
}

export function registerAgent(
  registry: DeployAgentRegistry,
  target: string,
  url: string,
  token?: string,
): DeployAgentRegistry {
  // Description:
  //     RegisterAgent.
  //
  // Inputs:
  //     registry: DeployAgentRegistry
  //         Caller-supplied registry.
  //     target: string
  //         Caller-supplied target.
  //     url: string
  //         Caller-supplied url.
  //     token?: string
  //         Caller-supplied token?.
  //
  // Outputs:
  //     result: DeployAgentRegistry
  //         Return value from `registerAgent`.
  //
  // Example:

  //     const result = registerAgent(registry, target, url, token?);

  if (!url.startsWith("http://") && !url.startsWith("https://")) {
    throw new Error(`deploy agent URL must start with http:// or https:// (got ${url})`);
  }
  const agents = registry.agents.filter((entry) => entry.target !== target);
  agents.push({ target, url, token });
  agents.sort((a, b) => a.target.localeCompare(b.target));
  return { agents };
}

export function lookupAgent(registry: DeployAgentRegistry, target: string): DeployAgentEntry | undefined {
  // Description:
  //     LookupAgent.
  //
  // Inputs:
  //     registry: DeployAgentRegistry
  //         Caller-supplied registry.
  //     target: string
  //         Caller-supplied target.
  //
  // Outputs:
  //     result: DeployAgentEntry | undefined
  //         Return value from `lookupAgent`.
  //
  // Example:

  //     const result = lookupAgent(registry, target);

  return registry.agents.find((entry) => entry.target === target);
}

async function agentFetch(
  entry: DeployAgentEntry,
  method: string,
  path: string,
  body?: string,
): Promise<Response> {
  // Description:
  //     AgentFetch.
  //
  // Inputs:
  //     entry: DeployAgentEntry
  //         Caller-supplied entry.
  //     method: string
  //         Caller-supplied method.
  //     path: string
  //         Caller-supplied path.
  //     body?: string
  //         Caller-supplied body?.
  //
  // Outputs:
  //     result: Promise<Response>
  //         Return value from `agentFetch`.
  //
  // Example:

  //     const result = agentFetch(entry, method, path, body?);

  const base = entry.url.replace(/\/$/, "");
  const headers: Record<string, string> = { Accept: "application/json" };
  if (body) headers["Content-Type"] = "application/json";
  if (entry.token) headers.Authorization = `Bearer ${entry.token}`;
  return remoteFetch(`${base}${path}`, { method, headers, body });
}

export async function agentHealth(entry: DeployAgentEntry): Promise<boolean> {
  // Description:
  //     AgentHealth.
  //
  // Inputs:
  //     entry: DeployAgentEntry
  //         Caller-supplied entry.
  //
  // Outputs:
  //     result: Promise<boolean>
  //         Return value from `agentHealth`.
  //
  // Example:

  //     const result = agentHealth(entry);

  const response = await agentFetch(entry, "GET", "/v1/health");
  if (!response.ok) return false;
  const body = (await response.json()) as { ok?: boolean };
  return body.ok === true;
}

export async function agentReadiness(
  entry: DeployAgentEntry,
  runtime = false,
  injectHealthFaults = false,
): Promise<{
  // Description:
  //     AgentReadiness.
  //
  // Inputs:
  //     entry: DeployAgentEntry
  //         Caller-supplied entry.
  //     runtime = false: input value
  //         Caller-supplied runtime = false.
  //     injectHealthFaults = false: input value
  //         Caller-supplied injectHealthFaults = false.
  //
  // Outputs:
  //     result: Promise<
  //         Return value from `agentReadiness`.
  //
  // Example:

 // const result = agentReadiness(entry, runtime = false, injectHealthFaults = false);
 ok: boolean; mission_ready?: boolean; readiness?: unknown }> {
  const query = new URLSearchParams();
  if (runtime) query.set("runtime", "true");
  if (injectHealthFaults) query.set("inject_health_faults", "true");
  const suffix = query.toString() ? `?${query.toString()}` : "";
  const response = await agentFetch(entry, "GET", `/v1/readiness${suffix}`);
  if (!response.ok) {
    throw new Error(`agent readiness HTTP ${response.status}`);
  }
  return (await response.json()) as { ok: boolean; mission_ready?: boolean; readiness?: unknown };
}

export async function agentUploadProgram(entry: DeployAgentEntry, program: string): Promise<void> {
  // Description:
  //     AgentUploadProgram.
  //
  // Inputs:
  //     entry: DeployAgentEntry
  //         Caller-supplied entry.
  //     program: string
  //         Caller-supplied program.
  //
  // Outputs:
  //     result: Promise<void>
  //         Return value from `agentUploadProgram`.
  //
  // Example:

  //     const result = agentUploadProgram(entry, program);

  const response = await agentFetch(entry, "POST", "/v1/program", JSON.stringify({ program }));
  if (!response.ok) {
    throw new Error(`agent program upload HTTP ${response.status}`);
  }
  const body = (await response.json()) as { ok?: boolean };
  if (!body.ok) {
    throw new Error("agent program upload failed");
  }
}

export async function agentStatus(entry: DeployAgentEntry): Promise<AgentStatusResponse> {
  // Description:
  //     AgentStatus.
  //
  // Inputs:
  //     entry: DeployAgentEntry
  //         Caller-supplied entry.
  //
  // Outputs:
  //     result: Promise<AgentStatusResponse>
  //         Return value from `agentStatus`.
  //
  // Example:

  //     const result = agentStatus(entry);

  const response = await agentFetch(entry, "GET", "/v1/status");
  if (!response.ok) {
    throw new Error(`agent status HTTP ${response.status}`);
  }
  const body = (await response.json()) as AgentStatusResponse & { current_version?: string; previous_version?: string };
  return {
    ok: body.ok,
    target: body.target,
    currentVersion: body.currentVersion ?? body.current_version ?? "",
    previousVersion: body.previousVersion ?? body.previous_version,
    healthy: body.healthy,
  };
}

export async function executeRemoteRollout(
  plan: DeployPlan,
  options: RolloutOptions,
  registry: DeployAgentRegistry,
  bundle: DeployArtifactBundle,
): Promise<RolloutResult> {
  // Description:
  //     ExecuteRemoteRollout.
  //
  // Inputs:
  //     plan: DeployPlan
  //         Caller-supplied plan.
  //     options: RolloutOptions
  //         Caller-supplied options.
  //     registry: DeployAgentRegistry
  //         Caller-supplied registry.
  //     bundle: DeployArtifactBundle
  //         Caller-supplied bundle.
  //
  // Outputs:
  //     result: Promise<RolloutResult>
  //         Return value from `executeRemoteRollout`.
  //
  // Example:

  //     const result = executeRemoteRollout(plan, options, registry, bundle);

  const local = planRollout(plan, options);
  if (options.dryRun) return local;

  const steps: RolloutStep[] = [];
  let success = true;
  for (const step of local.steps) {
    if (step.status === "skipped") {
      steps.push(step);
      continue;
    }
    const key = deployTargetKey(step.robotName, step.hardware);
    const agent = lookupAgent(registry, key);
    if (!agent) {
      steps.push({ ...step, status: "failed" });
      success = false;
      continue;
    }
    try {
      const response = await agentFetch(
        agent,
        "POST",
        "/v1/rollout",
        JSON.stringify({
          target: key,
          version: bundle.version,
          program: bundle.program,
          program_hash: bundle.programHash,
          assignments: bundle.assignments.map((assignment) => ({
            robot_name: assignment.robotName,
            hardware: assignment.hardware,
          })),
          certifications: bundle.certifications,
          certification_proof: certificationProofPayload(plan.certificationProof),
          artifact_signature: bundle.signature,
          artifact_public_key: bundle.publicKey,
        }),
      );
      const body = (await response.json()) as { ok?: boolean; version?: string };
      if (response.ok && body.ok) {
        steps.push({ ...step, status: "deployed" });
      } else {
        steps.push({ ...step, status: "failed", version: body.version ?? step.version });
        success = false;
      }
    } catch {
      steps.push({ ...step, status: "failed" });
      success = false;
    }
  }
  return {
    strategy: options.strategy,
    version: options.version,
    dryRun: false,
    steps,
    success,
  };
}

export async function executeRemoteRollback(
  plan: DeployPlan,
  registry: DeployAgentRegistry,
): Promise<RolloutResult> {
  // Description:
  //     ExecuteRemoteRollback.
  //
  // Inputs:
  //     plan: DeployPlan
  //         Caller-supplied plan.
  //     registry: DeployAgentRegistry
  //         Caller-supplied registry.
  //
  // Outputs:
  //     result: Promise<RolloutResult>
  //         Return value from `executeRemoteRollback`.
  //
  // Example:

  //     const result = executeRemoteRollback(plan, registry);

  const steps: RolloutStep[] = [];
  let success = false;
  for (const assignment of plan.assignments) {
    const key = deployTargetKey(assignment.robotName, assignment.hardware);
    const agent = lookupAgent(registry, key);
    const base: RolloutStep = {
      robotName: assignment.robotName,
      hardware: assignment.hardware,
      status: "skipped",
      version: "unknown",
      phasePercent: null,
    };
    if (!agent) {
      steps.push(base);
      continue;
    }
    try {
      const response = await agentFetch(agent, "POST", "/v1/rollback", JSON.stringify({ target: key }));
      const body = (await response.json()) as { ok?: boolean; version?: string };
      if (response.ok && body.ok) {
        success = true;
        steps.push({ ...base, status: "rolled_back", version: body.version ?? "unknown" });
      } else {
        steps.push({ ...base, status: "failed", version: body.version ?? "unknown" });
      }
    } catch {
      steps.push({ ...base, status: "failed" });
    }
  }
  return {
    strategy: "all",
    version: "rollback",
    dryRun: false,
    steps,
    success,
  };
}

export function applyRemoteRolloutToState(
  state: import("./deploy-service.js").DeployState,
  result: RolloutResult,
): void {
  // Description:
  //     ApplyRemoteRolloutToState.
  //
  // Inputs:
  //     state: import("./deploy-service.js").DeployState
  //         Caller-supplied state.
  //     result: RolloutResult
  //         Caller-supplied result.
  //
  // Outputs:
  //     None.
  //
  // Example:

  //     const result = applyRemoteRolloutToState(state, result);

  applyRollout(state, result);
}
